src/workers/event.rs

```diff
@@ -0,0 +1,545 @@
+            .await?;
+
+        // NOTE: We can return an error here and then nack because we'll just reupdate the host data
+        // with the exact same host heartbeat entry. There is no possibility of a duplicate
+        self.heartbeat_provider_update(lattice_id, host).await?;
+
+        // NOTE: We can return an error here and then nack because we'll just reupdate the host data
+        // with the exact same host heartbeat entry. There is no possibility of a duplicate
+        self.heartbeat_actor_update(lattice_id, host).await?;
+
+        Ok(())
+    }
+
+    #[instrument(level = "debug", skip(self, host), fields(host_id = %host.id))]
+    async fn handle_host_started(
+        &self,
+        lattice_id: &str,
+        host: &HostStarted,
+    ) -> anyhow::Result<()> {
+        debug!("Updating store with new host");
+        // New hosts have nothing running on them yet, so just drop it in the store
+        self.store
+            .store(lattice_id, host.id.clone(), Host::from(host))
+            .await
+            .map_err(anyhow::Error::from)
+    }
+
+    #[instrument(level = "debug", skip(self, host), fields(host_id = %host.id))]
+    async fn handle_host_stopped(
+        &self,
+        lattice_id: &str,
+        host: &HostStopped,
+    ) -> anyhow::Result<()> {
+        debug!("Handling host stopped event");
+        // NOTE(thomastaylor312): Generally to get a host stopped event, the host should have
+        // already sent a bunch of stop actor/provider events, but for correctness sake, we fetch
+        // the current host and make sure all the actors and providers are removed
+        trace!("Fetching current host data");
+        let current: Host = match self.store.get(lattice_id, &host.id).await? {
+            Some(h) => h,
+            None => {
+                debug!("Got host stopped event for a host we didn't have in the store");
+                return Ok(());
+            }
+        };
+
+        trace!("Fetching actors from store to remove stopped instances");
+        let all_actors = self.store.list::<Actor>(lattice_id).await?;
+
+        let actors_to_update = all_actors.into_iter().filter_map(|(id, mut actor)| {
+            if current.actors.contains_key(&id) {
+                actor.count.remove(&current.id);
+                Some((id, actor))
+            } else {
+                None
+            }
+        });
+        trace!("Storing updated actors in store");
+        self.store.store_many(lattice_id, actors_to_update).await?;
+
+        trace!("Fetching providers from store to remove stopped instances");
+        let mut all_providers = self.store.list::<Provider>(lattice_id).await?;
+
+        let providers_to_update = current.providers.into_iter().filter_map(|info| {
+            let key = crate::storage::provider_id(&info.public_key, &info.link_name);
+            let mut prov = all_providers.remove(&key)?;
+            prov.hosts.remove(&host.id).map(|_| (key, prov))
```

Comment:
I found this logic to be confusing, and I'm still not sure I grok it. Here's what I think we're doing:
- retrieve `all_providers` running in the lattice
- for each provider that was running on `current`, remove it from `all_providers`
  - question: why `remove` instead of `get` or `entry`? A provider with a unique `(pk, link)` ID can exist multiple times within the lattice (once per host). Calling `remove` caused me to infer we were purposefully removing from the lattice state _entirely_, which would be incorrect; but we're _not_ actually doing that, since we're not referring to `all_providers` after this loop, and since we're only processing one host ID here. So this seems okay, but unnecessarily confusing and prone to future bugs. I assume you did this to get a `Provider` instead of a `&Provider`, so you could call `prov.hosts.remove`. IMO this is an example where Rust's type system is incentivizing confusing code and I'd rather work around the types to write clearer code
- if we found a provider with the given key, try to remove the stopped host from its map. If we get a result, return a `(key, prov)` tuple as the new state we'll store. Otherwise return None since this entry doesn't need to be updated

Assuming I've understood this, I think something like this is more clear:
```rust
let mut maybe_prov = (*all_providers.get(&key)).clone();
match maybe_prov {
  Some(prov) => match prov.hosts.remove(&host.id) {
    Some(_) => (key, prov)
    None => None // provider wasn't running on this host
  }
  None => None // provider wasn't running in the lattice (maybe warn here? This seems unexpected)
}
```

Normally I prefer `map` to matching like this, but given the complexity here it might be worth calling out the individual cases

Reply:
Yes, you read the logic properly and I was trying to avoid the clone (which is why I was _removing_ from the `all_providers` map, which is owned data). We're probably just fine to clone here and I'll make a note that it was a purposeful tradeoff

Reply:
Yes, you read the logic properly and I was trying to avoid the clone (which is why I was _removing_ from the `all_providers` map, which is owned data). We're probably just fine to clone here and I'll make a note that it was a purposeful tradeoff

Reply:
Just a heads up that clippy didn't like that inner match, so I stuck with the `remove().map()`